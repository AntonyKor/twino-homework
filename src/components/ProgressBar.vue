<template>
  <div class="progress-bar" :style="barStyles">
    <div
      v-for="n in groupsNumber"
      :key="n"
      :class="{
        filled: n - 1 < currentGroup,
        active: n - 1 === currentGroup,
      }"
      class="group"
    >
      <div
        v-if="n - 1 === currentGroup"
        :style="currentGroupProgressStyles"
        class="filled"
      ></div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    groupsNumber: {
      type: Number,
      required: true,
    },
    currentGroup: {
      type: Number,
      required: true,
    },
    currentGroupProgress: {
      type: Number,
      required: true,
    },
  },

  computed: {
    barStyles() {
      return { gridTemplateColumns: `repeat(${this.groupsNumber}, 1fr)` };
    },

    currentGroupProgressStyles() {
      return { width: `${this.currentGroupProgress}%` };
    },
  },
};
</script>

<style scoped lang="scss">
.progress-bar {
  display: grid;
  grid-gap: 12px;
}

.group {
  background-color: #eef1f8;
  height: 6px;
}

.filled {
  background-color: #81c784;
}

.active {
  position: relative;
  background-color: #d9eeda;

  & > div {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
  }
}
</style>
