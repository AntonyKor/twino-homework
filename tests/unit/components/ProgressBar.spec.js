import { shallowMount } from "@vue/test-utils";
import ProgressBar from "@/components/ProgressBar.vue";

describe("ProgressBar.vue", () => {
  it("should match snapshot", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 3, currentGroup: 1, currentGroupProgress: 35 },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render number of divs as in groupsNumber prop", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 3, currentGroup: 1, currentGroupProgress: 35 },
    });

    expect(wrapper.findAll(".group")).toHaveLength(3);
  });

  it("should apply grid-template-columns rule to container", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 3, currentGroup: 1, currentGroupProgress: 35 },
    });

    expect(wrapper.find(".progress-bar").attributes("style")).toEqual(
      "grid-template-columns: repeat(3, 1fr);"
    );
  });

  it("should give filled class for groups before current", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 5, currentGroup: 2, currentGroupProgress: 35 },
    });

    const groups = wrapper.findAll(".group");
    expect(groups.at(0).classes("filled")).toBeTruthy();
    expect(groups.at(1).classes("filled")).toBeTruthy();
    expect(groups.at(2).classes("filled")).toBeFalsy();
    expect(groups.at(3).classes("filled")).toBeFalsy();
    expect(groups.at(4).classes("filled")).toBeFalsy();
  });

  it("should give active class for current group", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 5, currentGroup: 2, currentGroupProgress: 35 },
    });

    const groups = wrapper.findAll(".group");
    expect(groups.at(0).classes("active")).toBeFalsy();
    expect(groups.at(1).classes("active")).toBeFalsy();
    expect(groups.at(2).classes("active")).toBeTruthy();
    expect(groups.at(3).classes("active")).toBeFalsy();
    expect(groups.at(4).classes("active")).toBeFalsy();
  });

  it("should render child div in current group", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 5, currentGroup: 2, currentGroupProgress: 35 },
    });

    const groups = wrapper.findAll(".group");
    expect(groups.at(0).isEmpty()).toBeTruthy();
    expect(groups.at(1).isEmpty()).toBeTruthy();
    expect(groups.at(2).isEmpty()).toBeFalsy();
    expect(groups.at(3).isEmpty()).toBeTruthy();
    expect(groups.at(4).isEmpty()).toBeTruthy();
  });

  it("should apply width on child div in current group", () => {
    const wrapper = shallowMount(ProgressBar, {
      propsData: { groupsNumber: 5, currentGroup: 2, currentGroupProgress: 35 },
    });

    expect(wrapper.find(".active div").attributes("style")).toEqual(
      "width: 35%;"
    );
  });
});
