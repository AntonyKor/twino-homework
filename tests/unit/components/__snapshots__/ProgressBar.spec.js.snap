// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar.vue should match snapshot 1`] = `
<div
  class="progress-bar"
  style="grid-template-columns: repeat(3, 1fr);"
>
  <div
    class="group filled"
  >
    <!---->
  </div>
  <div
    class="group active"
  >
    <div
      class="filled"
      style="width: 35%;"
    />
  </div>
  <div
    class="group"
  >
    <!---->
  </div>
</div>
`;
